addons/mail/models/discuss/discuss_channel.py

```diff
@@ -915,13 +915,20 @@ def channel_get(self, partners_to, pin=True):
             channel._broadcast(partners_to)
         return channel
 
-    def channel_fold(self, state=None, state_count=0):
+    def _channel_fold(self, state=None, state_count=0):
         """ Update the fold_state of the given session. In order to syncronize web browser
-            tabs, the change will be broadcast to themselves (the current user channel).
+            tabs, the change will be broadcast to themselves (the current persona channel).
             Note: the user need to be logged
-            :param state : the new status of the session for the current user.
+            :param state : the new status of the session for the current persona.
         """
-        domain = [('partner_id', '=', self.env.user.partner_id.id), ('channel_id', 'in', self.ids)]
+        current_partner, current_guest = self.env["res.partner"]._get_current_persona()
+        domain = expression.AND([
+            [("channel_id", "in", self.ids)],
+            expression.OR([
+                [("partner_id", "=", current_partner.id)] if current_partner else [],
+                [("guest_id", "=", current_guest.id)] if current_guest else [],
+            ])
+        ])
```

Comment:
cleanup: `_channel_fold` should be moved to channel member so that we don't need to do this search again
controller can directly use `member = request.env["discuss.channel.member"].search([("channel_id", "=", channel_id), ("is_self", "=", True)])`